build(yarn)!: migrate to Yarn v4 (Berry) with Corepack; update CI/Docker and workspace tooling

nickhsine opened the pull request:
## Description
- Yarn v4 migration
    - Set packageManager: "yarn@^4.9.4" at root
    - Configure .yarnrc.yml with nodeLinker: node-modules
    - Regenerate lockfile for Yarn v4
- CI/Docker updates
    - Cloud Build: copy .yarnrc.yml into subpackage context before docker build
    - Dockerfiles (api-gateway, frontend, cms, cronjob, image-resizer):
        - RUN corepack enable in stages
        - copy .yarnrc.yml with manifests
        - switch installs to yarn install --immutable
- Workspace scripts
    - Replace deprecated workspaces run with workspaces foreach -A run ... for lint, format, and type-check
    - Keeps Husky pre-push working under Yarn v4
- Dependency alignment
    - Move build-time tooling (babel/webpack) from root into packages that use them
    - Add missing peer/dev dependencies to satisfy Yarn v4 validation (react/react-dom, keystone peers, babel presets/plugins)
    - Use yarn exec for package-local CLIs in Makefiles
- Documentation
    - README: Corepack activation, immutable installs, Docker/Corepack notes

## Breaking changes
- Requires Node 18/20 with Corepack enabled
- CI should use yarn install --immutable
- Local initial install: run corepack enable && yarn install --no-immutable once to write the v4 lockfile

## Validation
- Local: yarn type-check, yarn lint:check, and package builds run using per-package toolchains
- Docker: builds succeed using subpackage context + Corepack
- Cloud Build: uses immutable installs and Yarn v4 config in each subpackage context

nickhsine committed:
build(root)!: switch to Yarn v4 (Berry) with Corepack

- Migrate from Yarn v1 to Yarn v4
- Set packageManager: "yarn@^4.9.4" in root package.json
- Keep node_modules via .yarnrc.yml (nodeLinker: node-modules)
- Note: breaking change for local/CI environments — requires Node 18/20 and corepack enable; CI should use yarn install --immutable

nickhsine committed:
chore(deps): align workspace deps for Yarn v4

- root: trim devDependencies; pin packageManager to yarn@^4.3.1
- cms: add react/react-dom deps; make SKIP_BUILD_DEPS optional in postinstall
- core: add react/react-dom as peers; add babel devDeps for local builds
- draft-editor: add devDeps (babel/webpack) for builds
- draft-renderer: move build deps to devDeps; add babel devDeps
- frontend: add @types/react devDep
- routing-ui: add react/react-dom devDeps to satisfy peer checks

nickhsine committed:
build(ci): adopt Corepack + Yarn v4 across subpackages

- cloudbuild: copy .yarnrc.yml into packages/$_TARGET_PACKAGE before docker build to apply Yarn v4 settings
- Dockerfiles (api-gateway, frontend, cms, cronjob, image-resizer, image-resizer/all):
  - add RUN corepack enable in build/runtime stages
  - copy .yarnrc.yml alongside manifests
  - switch installs to yarn install --immutable
- cms Dockerfile: keep SKIP_BUILD_DEPS=true while using immutable install
- Makefiles (core, draft-editor, draft-renderer): use yarn exec to execute babel/webpack binaries under Yarn v4

nickhsine committed:
chore(lockfile): regenerate Yarn v4 lockfile

- Regenerated yarn.lock after migrating to Yarn v4 (Berry)

nickhsine committed:
docs(root): document Yarn v4 + Corepack workflow

- Add Yarn v4 (Berry) notes to README:
  - Corepack activation and packageManager pin
  - node-modules linker and immutable installs
  - Subpackage Docker builds copy .yarnrc.yml
- Clarify CI/Docker usage of yarn install --immutable and local corepack enable

nickhsine committed:
chore(scripts): replace deprecated workspaces run with workspaces foreach

- Root scripts now use Yarn v4 syntax:
  - format: yarn workspaces foreach -A run prettier --write|--check
  - lint: yarn workspaces foreach -A run lint:check (and --fix)
  - type-check: yarn workspaces foreach -A run type-check
- Fixes husky pre-push error (“workspaces run not found”) under Yarn v4
- -A/--all ensures every workspace participates, matching previous behavior

nickhsine commented:
關於 validation 的部分，我這邊目前的規劃是：
1. 新增 yarnv4 branch, based from ndx branch
2. 新增 cloud build trigger：cms 和 frontend trigger
3. 透過 PR 中的 cloudbuild.yaml 來部署 yarnv4 版本的 code base
4. 在 cloud run 上測試 cms 和 frontend ，看會不會有異常狀況
5. 確定沒問題，再把 yarnv4 merge 回 ndx branch

ericsen-tsai commented:
> 關於 validation 的部分，我這邊目前的規劃是：
> 
> 1. 新增 yarnv4 branch, based from ndx branch
> 2. 新增 cloud build trigger：cms 和 frontend trigger
> 3. 透過 PR 中的 cloudbuild.yaml 來部署 yarnv4 版本的 code base
> 4. 在 cloud run 上測試 cms 和 frontend ，看會不會有異常狀況
> 5. 確定沒問題，再把 yarnv4 merge 回 ndx branch

OK, sounds good! go go~

ericsen-tsai commented:
suggestion:
1. You can use wildcard to include all packages inside the folder now

```json
{
  "workspaces": [
    "packages/*"
  ]
}
```

reference: https://yarnpkg.com/features/workspaces#how-are-workspaces-declared

2. In v4, the package version across package can now use workspace protocol

```json
{
  "dependencies": {
    "@my-org/utils": "workspace:^"
  }
}
```

reference: https://yarnpkg.com/features/workspaces#how-are-workspaces-declared

ericsen-tsai commented on README.md:
```diff
@@ -21,7 +21,14 @@ Before modifying sub-packages' source codes, make sure you install dependencies
 
 ### Installation
-`yarn install`
+- Corepack (Yarn v4): ensure Node 18/20 and run `corepack enable` once on your machine.
```
question:
Since this repo uses Node 20, maybe we just need to mention this version?

nickhsine replied:
Done in [docs(root): update README.md](https://github.com/kids-reporter/kids-reporter-monorepo/pull/793/commits/ad6040cab668b8867e247e4adf3f0da0ac4dcbd6)

nickhsine committed:
chore(root): use workspaces glob packages/*

Collapse explicit subpackage listing into a wildcard to auto-include new
packages.

nickhsine commented:
> suggestion:
> 
> 1. You can use wildcard to include all packages inside the folder now
> 
> ```json
> {
>   "workspaces": [
>     "packages/*"
>   ]
> }
> ```

Done in [chore(root): use wildcard "packages/*" to replace subpkg listing](https://github.com/kids-reporter/kids-reporter-monorepo/pull/793/commits/80df99e06ab249dab23d1dbc2407172529fb05b7)


> 
> reference: https://yarnpkg.com/features/workspaces#how-are-workspaces-declared
> 
> 2. In v4, the package version across package can now use workspace protocol
> 
> ```json
> {
>   "dependencies": {
>     "@my-org/utils": "workspace:^"
>   }
> }
> ```
> 
> reference: https://yarnpkg.com/features/workspaces#how-are-workspaces-declared

因為 packages/frontend 和 packages/cms 不會 publish NPM package，
而且在 Cloud Build 階段，並不是使用 yarn workspace 的方式去 install pkgs，所以現階段使用 workspace protocol 的效益不大。
經討論，暫時先不調整目前的寫法。

nickhsine commented on packages/cms/package.json:
```diff
@@ -7,8 +7,8 @@
     "build": "keystone build",
-    "postinstall": "SKIP_BUILD_DEPS=\"$SKIP_BUILD_DEPS\" make postinstall",
     "lint:check": "eslint . --ext .ts,.tsx,.js,.jsx",
+    "postinstall": "SKIP_BUILD_DEPS=\"${SKIP_BUILD_DEPS:-}\" make postinstall",
```
Done in [chore(cms): update postinstall script to work correctly in all environments (e.g., Windows])(https://github.com/kids-reporter/kids-reporter-monorepo/pull/793/commits/4987e465a90151fb88e507c99130cd0cf603a6dc)

nickhsine commented:
我先用目前的版本 merge 到 yarnv4 branch 去測試，如果還有需要修改或是討論的地方，歡迎留言，謝謝。

nickhsine committed:
chore(de,dr): update Makefile build-lib target

Remove special views handling and .views.babelrc usage.

nickhsine commented:
> > 關於 validation 的部分，我這邊目前的規劃是：
> > 
> > 1. 新增 yarnv4 branch, based from ndx branch
> > 2. 新增 cloud build trigger：cms 和 frontend trigger
> > 3. 透過 PR 中的 cloudbuild.yaml 來部署 yarnv4 版本的 code base
> > 4. 在 cloud run 上測試 cms 和 frontend ，看會不會有異常狀況
> > 5. 確定沒問題，再把 yarnv4 merge 回 ndx branch
> 
> OK, sounds good! go go~

frontend testing URL: https://yarnv4-kids.twreporter.org
cms testing URL:  https://yarnv4-kids-cms.twreporter.org 

目前看起來沒有什麼問題。
